src/oid4vp/holder.rs

```diff
@@ -0,0 +1,511 @@
+use super::error::OID4VPError;
+use super::permission_request::*;
+use crate::common::*;
+use crate::credential::*;
+use crate::vdc_collection::VdcCollection;
+
+use std::sync::Arc;
+
+use futures::StreamExt;
+use oid4vp::core::authorization_request::parameters::ClientIdScheme;
+use oid4vp::core::credential_format::ClaimFormatDesignation;
+use oid4vp::core::presentation_definition::PresentationDefinition;
+use oid4vp::{
+    core::{
+        authorization_request::{
+            parameters::ResponseMode,
+            verification::{did::verify_with_resolver, RequestVerifier},
+            AuthorizationRequestObject,
+        },
+        metadata::WalletMetadata,
+        presentation_submission::{DescriptorMap, PresentationSubmission},
+        response::{parameters::VpToken, AuthorizationResponse, UnencodedAuthorizationResponse},
+    },
+    // verifier::request_signer::RequestSigner,
+    wallet::Wallet as OID4VPWallet,
+};
+// use ssi::claims::jws::JwsSigner;
+use ssi::dids::DIDWeb;
+use ssi::dids::VerificationMethodDIDResolver;
+use ssi::prelude::AnyJwkMethod;
+// use ssi::JWK;
+use tokio::sync::RwLock;
+use uniffi::deps::{anyhow, log};
+
+/// A Holder is an entity that possesses one or more Verifiable Credentials.
+/// The Holder is typically the subject of the credentials, but not always.
+/// The Holder has the ability to generate Verifiable Presentations from
+/// these credentials and share them with Verifiers.
+#[derive(Debug, uniffi::Object)]
+pub struct Holder {
+    /// An atomic reference to the VDC collection.
+    pub(crate) vdc_collection: Option<Arc<VdcCollection>>,
+
+    /// Metadata about the holder.
+    pub(crate) metadata: WalletMetadata,
+
+    /// HTTP Request Client
+    pub(crate) client: oid4vp::core::util::ReqwestClient,
+
+    /// A list of trusted DIDs.
+    pub(crate) trusted_dids: Vec<String>,
+
+    // /// A request signer callback interface to handle
+    // /// cryptographic signing with native libraries.
+    // pub(crate) request_signer: Option<Arc<dyn RequestSignerInterface>>,
+    /// Provide optional credentials to the holder instance.
+    pub(crate) provided_credentials: Option<Vec<Arc<Credential>>>,
+
+    /// Holder a reference to the current authorization request object.
+    pub(crate) authorization_request: RwLock<Option<AuthorizationRequestObject>>,
```

Comment:
To avoid race conditions I would strongly recommend removing the authorization request from the Holder struct, and instead return a "Session" object from the `authorization_request` function.

Reply:
I don't imagine we'll have a race condition here. This is more for providing inner immutability when accepting the authorization request.

Also, with keeping a session, we're not guaranteed from race conditions there either, as it would presumably also be wrapped in something like an `RwLock` to keep track of the session on the holder side.

I think this is okay to merge as is and we can address any faults with it if we encounter them.

Reply:
I've updated the code to pass ownership of the authorization request object to the permission request object to use in the permission response. 

This removes the internal state of the authorization request and the `RwLock`. This is more closely related to the `Session` style object and removes the chance of race conditions.